### ISSUE TITLE
# Concept: Pre-Warmed Worker Thread for `action_graph::Trigger`

### ISSUE THREAD

**Issue by danieldube** [danieldube/action_graph#28]
**Status:** Proposal / discussion draft
**Audience:** `action_graph` maintainers
**Scope:** `include/action_graph/global_timer/trigger.h`, `src/action_graph/global_timer/trigger.cpp`
**Author's role:** Consumer of `action_graph` as a library, reporting an observed limitation and proposing a backward-compatible fix

---

## 1. Motivation

`action_graph::GlobalTimer` schedules periodic `Action`s and fires them by calling `Trigger::TriggerAsynchronously()`. The current implementation of `Trigger` (see `trigger.cpp`) creates a **brand-new, detached `std::thread`** on every single firing:

```cpp
void Trigger::TriggerAsynchronously() {
  if (is_running_.exchange(true)) {
    return;
  }
  std::thread([this]() {
    callback_();
    is_running_ = false;
  }).detach();
}
```

This is simple and correct, but it means the **cost and latency of `pthread_create`/`clone(2)` (and the corresponding thread teardown) sits directly on the critical path of every periodic action**, for every action in the graph, for every cycle.

For low-frequency actions (say, periods of hundreds of milliseconds or more) this overhead is negligible. However, for use cases with:

- many concurrently scheduled triggers in the same `GlobalTimer` (tens of independent periodic actions),
- short periods (single-digit to low-double-digit milliseconds),
- and/or execution on a host under general CPU/scheduler load (shared CI runners, containerized environments, oversubscribed cores),

the **thread-creation and teardown latency is not constant** — it depends on kernel scheduler state, memory allocator contention, and the number of threads being concurrently created/destroyed elsewhere in the process. Under load, this latency can occasionally spike from the usual tens of microseconds up to several milliseconds. When many triggers share the same process, this creates cumulative contention (thread creation is itself a contended kernel resource), meaning the *tail* latency degrades faster than linearly with the number of concurrently active triggers.

**Consequence:** consumers who use `GlobalTimer` for tight periodic scheduling (e.g., sub-50 ms periods, with a dozen or more concurrently registered triggers) can observe occasional, hard-to-reproduce timing violations that are not caused by the user-supplied callback's own execution time, but purely by the fixed and variable overhead of creating a new OS thread per firing. This is difficult for library consumers to diagnose because it manifests as "callback took too long" or "trigger missed its period," even though the callback body itself executes quickly — the delay is hidden inside `std::thread` construction and destruction, which is opaque to the caller.

This proposal asks: **can `Trigger` avoid re-creating an OS thread on every firing**, while preserving all existing observable behavior?

---

## 2. Goals and Non-Goals

**Goals**

1. Eliminate repeated OS thread creation/destruction from the steady-state firing path of `Trigger`.
2. Preserve `Trigger`'s existing public API and behavior exactly — this must be a drop-in internal change with **zero required changes** for existing callers of `Trigger` or `GlobalTimer`.
3. Preserve existing semantics:
   - A `Trigger` is created with a callback that returns `void` and takes no arguments.
   - Calling `TriggerAsynchronously()` while a previous invocation of the same trigger is still running is a **no-op** (never queues, never blocks the caller).
   - `WaitUntilTriggerIsFinished()` blocks the calling thread until the currently running invocation (if any) completes.
   - The destructor waits for any in-flight invocation to finish before releasing resources.
4. Keep the change small, self-contained, and reviewable (single class, no new public headers/types required).

**Non-Goals** (explicitly out of scope for this proposal; noted as possible follow-ups)

- Pooling/sharing worker threads *across multiple* `Trigger` instances (would change `GlobalTimer`'s architecture and thread-count characteristics; a larger, separate design discussion).
- Making the dispatch strategy pluggable/configurable (a Strategy pattern could be layered on top of this change later, but is not necessary to solve the core problem).
- Changing `GlobalTimer`'s polling loop or scheduling algorithm.

---

## 3. Proposed Design

Replace the "one-shot detached thread per firing" implementation with a **persistent worker thread owned by the `Trigger` instance itself**, created once at `Trigger` construction and reused for every subsequent firing until the `Trigger` is destroyed. Firing becomes a lightweight condition-variable notification instead of a thread creation.

### 3.1 Public interface — unchanged

```cpp
class Trigger {
 public:
  explicit Trigger(std::function<void()> callback);

  Trigger(const Trigger &) = delete;
  Trigger(Trigger &&other) noexcept;
  Trigger &operator=(const Trigger &) = delete;
  Trigger &operator=(Trigger &&) = delete;

  ~Trigger();

  void TriggerAsynchronously();
  void WaitUntilTriggerIsFinished() const;

 private:
  // internals change, signature does not
};
```

No consumer-visible type, header, or method signature changes. `GlobalTimer` requires **no modification**.

### 3.2 Internal implementation sketch

```cpp
class Trigger {
 public:
  explicit Trigger(std::function<void()> callback)
      : callback_(std::move(callback)),
        worker_thread_([this]() { WorkerLoop(); }) {}

  ~Trigger() {
    {
      std::lock_guard<std::mutex> lock(mutex_);
      stop_requested_ = true;
    }
    condition_.notify_all();
    if (worker_thread_.joinable()) {
      worker_thread_.join();
    }
  }

  void TriggerAsynchronously() {
    std::lock_guard<std::mutex> lock(mutex_);
    if (is_running_) {
      return;  // preserves existing "skip if already running" behavior
    }
    is_running_ = true;
    fire_requested_ = true;
    condition_.notify_all();
  }

  void WaitUntilTriggerIsFinished() const {
    std::unique_lock<std::mutex> lock(mutex_);
    condition_.wait(lock, [this] { return !is_running_; });
  }

 private:
  void WorkerLoop() {
    std::unique_lock<std::mutex> lock(mutex_);
    while (true) {
      condition_.wait(lock, [this] {
        return fire_requested_ || stop_requested_;
      });
      if (stop_requested_ && !fire_requested_) {
        return;
      }
      fire_requested_ = false;
      lock.unlock();
      callback_();
      lock.lock();
      is_running_ = false;
      condition_.notify_all();
    }
  }

  std::function<void()> callback_;
  mutable std::mutex mutex_;
  std::condition_variable condition_;
  bool is_running_{false};
  bool fire_requested_{false};
  bool stop_requested_{false};
  std::thread worker_thread_;
};
```

Notes on the sketch (to be refined during implementation/review, not final code):

- `is_running_` and `WaitUntilTriggerIsFinished` move from `std::atomic<bool>` + busy-yield polling (current implementation) to mutex/condition-variable-guarded state. This is itself an improvement: the current `WaitUntilTriggerIsFinished()` busy-spins (`while (is_running_) std::this_thread::yield();`), burning CPU; the proposed version blocks efficiently.
- Move-construction (`Trigger(Trigger&& other)`) needs care: a `std::thread` and a `std::mutex`/`std::condition_variable` are not trivially movable together while the worker thread is running. The simplest correct option is to **disallow move once the worker thread has started**, or to restrict `Trigger` to be constructed in-place (e.g., via `std::vector::emplace_back`, which is how `GlobalTimer::ScheduledTrigger` already constructs it — see `global_timer.h`, `schedule_.emplace_back(...)`). This should be confirmed against all current call sites before finalizing the design; if move-after-construction-with-running-thread is required by any caller, the design must account for it explicitly (e.g., a two-phase "start worker on first fire" approach instead of starting it in the constructor).

### 3.3 Behavioral equivalence checklist

| Behavior | Current (detached thread per fire) | Proposed (persistent worker) |
|---|---|---|
| Re-entrant fire while running is ignored | Yes (`is_running_.exchange`) | Yes (`if (is_running_) return;`) |
| `WaitUntilTriggerIsFinished()` blocks until current run completes | Yes (busy-poll) | Yes (condition-variable wait) |
| Destructor waits for in-flight callback before returning | Yes (via `WaitUntilTriggerIsFinished()` in destructor) | Yes (join waits for worker to observe stop and exit; in-flight callback still completes because stop is only processed between firings) |
| Callback runs on a dedicated thread, not the timer's own thread | Yes | Yes |
| Number of OS threads per `Trigger` over its lifetime | 1 per firing (created & destroyed each time) | 1 total (created once, reused) |

---

### Benefits

- **Removes thread-creation/teardown latency from the firing critical path**, converting a variable, contention-sensitive cost (kernel thread creation) into a fixed, low, and predictable cost (futex-based condition-variable wake, typically low single-digit microseconds).
- **Reduces total OS/kernel load** in processes with many concurrently firing triggers, since thread objects are no longer continuously allocated and freed.
- **Replaces a busy-wait** (`WaitUntilTriggerIsFinished`'s `yield()` loop) with a blocking wait, reducing wasted CPU cycles — a secondary, independent improvement.
- **No API/behavior change** for existing consumers; purely an internal implementation improvement.

### Costs / Risks

- **Each `Trigger` now holds a thread for its entire lifetime**, rather than only while a callback is executing. For graphs with a very large number of triggers, this means a correspondingly large number of *idle, blocked* threads exist for the lifetime of the `GlobalTimer`. This is generally cheap (a blocked thread mostly costs kernel stack memory, not CPU), but it is a different resource profile than today's "threads exist only transiently" model, and should be called out in documentation.
- **Slightly more internal complexity**: mutex + condition variable + explicit state machine, versus an atomic flag + `std::thread(...).detach()`. This is a reasonable and idiomatic trade for correctness and performance, but increases the surface for subtle bugs (e.g., spurious wakeups, missed notifications, shutdown races) that need dedicated test coverage.
- **Move-semantics constraint** (see §3.2): if `Trigger` must remain movable after its worker thread has started, the implementation is more involved (e.g., `unique_ptr`-wrapped internal state so the `Trigger` object itself is trivially movable while the worker thread only ever touches the pointed-to state). This needs to be resolved based on real usage patterns in `GlobalTimer` and any other current callers.
- **No behavioral change means no fix for graphs with very many triggers and few CPU cores**: this proposal keeps a strict 1:1 relationship between triggers and worker threads. It does not address scenarios where the *number* of concurrently blocked/running threads itself becomes a scaling concern (e.g., hundreds of triggers). That would require a shared thread-pool design, which is a larger, separate architectural change and is intentionally out of scope here.

### Why this approach, and not alternatives

- A shared thread-pool inside `GlobalTimer` would solve both the creation-latency problem *and* the thread-count-scaling problem, but is a materially larger change (new queue/pool abstraction, changed `GlobalTimer` constructor, re-verification of per-trigger overlap-guard semantics across shared workers). It is better proposed and reviewed as its own, separate change once/if this smaller fix is accepted and the thread-count concern becomes relevant in practice.
- A pluggable dispatch-strategy abstraction (interface + multiple implementations) would let consumers choose between today's behavior and pre-warmed dispatch, but adds new public API surface and two code paths to maintain, for a benefit (opt-out capability) that may not be needed if the new behavior is a strict improvement with no observable downside for existing use cases.
- This proposal was chosen because it is the **smallest, most self-contained, purely-internal change** that fully removes the identified overhead while keeping the review surface, risk, and API impact minimal.

---

## 5. Further Points to Cover Before This Becomes a Pull Request

The following items should be addressed, ideally in discussion with maintainers, before or during implementation:

1. **Confirm move-semantics requirements.** Audit all current usages of `Trigger` (within `action_graph` and, if visible, any known downstream consumers) to determine whether a `Trigger` is ever moved *after* it has started firing, or only constructed in-place (e.g., via `emplace_back` into a container) and never moved thereafter. This determines whether the simple sketch in §3.2 is sufficient or whether an indirection (e.g., `pimpl`) is required for correct move semantics.
2. **Shutdown ordering.** Specify precisely what happens if `TriggerAsynchronously()` is called concurrently with destruction (should already be prevented by existing caller contracts, e.g. `GlobalTimer` locking `schedule_mutex_` around both trigger firing and timer-thread shutdown, but should be explicitly documented and tested for `Trigger` in isolation too).
3. **Spurious wakeup and missed-notification correctness.** Add explicit unit tests for:
   - Firing while idle (baseline behavior).
   - Firing while already running (must be ignored, not queued).
   - Calling `WaitUntilTriggerIsFinished()` before any fire has occurred (must return immediately).
   - Destruction while a callback is in flight (must block until completion, not abandon the thread).
   - Rapid repeated fire/wait cycles, to catch potential lost-wakeup races under `-fsanitize=thread`.
4. **Benchmark evidence.** Provide a micro-benchmark comparing per-fire latency of the current detached-thread implementation vs. the proposed persistent-worker implementation, ideally both on an idle machine and under artificial CPU/thread-creation contention (e.g., spawning background thread churn concurrently), to give maintainers concrete, reproducible numbers rather than only qualitative reasoning. This should be included in the PR description, not just this concept document.
5. **Resource-usage documentation update.** Update `Trigger`'s class documentation (and any relevant README/architecture notes) to state that a `Trigger` now owns one live OS thread for its entire lifetime rather than only during callback execution, so users instantiating very large numbers of `Trigger`s can make an informed decision.
6. **Thread naming / observability.** Consider setting a descriptive thread name (e.g., via `pthread_setname_np` on POSIX) for the persistent worker thread, to aid debugging/profiling (`top -H`, `perf`, gdb thread listings) — this was not possible/meaningful with the previous anonymous, short-lived detached threads, and is a natural improvement to include alongside this change.
7. **Sanitizer and platform coverage.** Ensure the change is validated under ThreadSanitizer and on all platforms the library currently supports/tests (the existing code is portable `std::thread`/`std::atomic`; the proposed code uses only `std::mutex`/`std::condition_variable`/`std::thread`, all standard and equally portable, but this should still be exercised in CI before merge).
8. **Changelog / versioning.** Since this changes internal resource-lifetime behavior (threads now live longer, see item 5) even though the API is unchanged, it may warrant a note in the project's changelog and, depending on the project's versioning policy, a minor version bump rather than being treated as a pure patch/bugfix.
9. **Scope confirmation with maintainers before implementation.** Given `Trigger`/`GlobalTimer` are small, central classes in this library, it is worth confirming with maintainers upfront (e.g., via an issue discussion) that this is the preferred direction, versus a shared-pool or pluggable-strategy design, before investing in a full implementation and test suite — to avoid rework if maintainers prefer a different long-term architecture for this concern.

---

## 6. Summary

The current `Trigger` implementation creates a new OS thread for every single firing. For consumers scheduling many short-period, concurrently-registered triggers, this puts a variable, contention-sensitive kernel operation (thread creation/teardown) directly on the timing-critical path, which can occasionally consume a large fraction of an otherwise generous period budget. This proposal replaces the per-firing detached thread with a persistent worker thread owned by each `Trigger`, created once and reused for the trigger's entire lifetime, communicating via a condition variable instead of thread creation. The change is designed to be a strictly internal, behavior-preserving improvement — no public API changes, no required changes in `GlobalTimer` or any other caller — while removing the dominant source of firing-latency variance. Remaining open items (move-semantics audit, benchmark evidence, expanded test coverage, and documentation of the new thread-lifetime model) are listed above as prerequisites for a maintainer-ready pull request.
